plugins/jvm-example/src/main/java/org/elasticsearch/plugin/example/JvmExamplePlugin.java

```diff
@@ -111,5 +117,16 @@ public Settings additionalSettings() {
     public void onModule(RepositoriesModule repositoriesModule) {
     }
 
-
+    /**
+     * Module decalaring some example configuration and a _cat action that uses
+     * it.
+     */
+    public static class ConfiguredExampleModule extends AbstractModule {
+        @Override
+        protected void configure() {
+          bind(ExamplePluginConfiguration.class).asEagerSingleton();
+          Multibinder<AbstractCatAction> catActionMultibinder = Multibinder.newSetBinder(binder(), AbstractCatAction.class);
```

Comment:
> The less our api relies on it, the easier it could be in the future to switch to something else.

I don't buy that argument.

> registration methods documented

I'd buy that argument if we bothered doing it. As a plugin developer the only chance I had of actually getting anything done was to read Elasticsearch's source and figure out where to insert myself.

> Having well defined extension points and effectively limiting the extensibility of es in genera

But we don't have well defined extension points.

> helps us make sure plugins cannot break things (as they'll be restricted to what we allow them to do)

I've developed plenty of plugins and I couldn't have done it without multibinder because Elasticsearch didn't expose all the right places so it'd work with onModule. I'd have had to fork Elasticsearch.

Besides that, why are we worried about plugin developers writing plugins that break things? If they do that its their problem.



I went and started doing this with onModule - and it requires so much more ceremony!

How is it less work to document the things you can bind then it is to document the Modules that you can use with onModule? I just blundered into the issue with the RestActionModule not being onModule-able.

I'm grumpy because I'm jet lagged - but I think I have a point here if you can read it around all the grump.